processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -468,39 +479,71 @@ public static void convertV8toV9(File v8Dir, File v9Dir) throws IOException
             }
 
             final VSizeIndexed finalMultiValCol = multiValCol;
-            singleValCol = VSizeIndexedInts.fromList(
-                new AbstractList<Integer>()
-                {
-                  @Override
-                  public Integer get(int index)
-                  {
-                    final VSizeIndexedInts ints = finalMultiValCol.get(index);
-                    return ints.size() == 0 ? 0 : ints.get(0) + (bumpedDictionary ? 1 : 0);
-                  }
-
-                  @Override
-                  public int size()
-                  {
-                    return finalMultiValCol.size();
-                  }
-                },
-                dictionary.size()
-            );
+            singleValCol = new AbstractList<Integer>()
+            {
+              @Override
+              public Integer get(int index)
+              {
+                final VSizeIndexedInts ints = finalMultiValCol.get(index);
+                return ints.size() == 0 ? 0 : ints.get(0) + (bumpedDictionary ? 1 : 0);
+              }
+
+              @Override
+              public int size()
+              {
+                return finalMultiValCol.size();
+              }
+            };
+
             multiValCol = null;
           } else {
             builder.setHasMultipleValues(true);
           }
 
-          builder.addSerde(
-              new DictionaryEncodedColumnPartSerde(
-                  dictionary,
-                  singleValCol,
-                  multiValCol,
-                  bitmapSerdeFactory,
-                  bitmaps,
-                  spatialIndex
-              )
-          );
+          final CompressedObjectStrategy.CompressionStrategy compression = indexSpec.getDimensionCompression();
+
+          if (singleValCol != null) {
+            if (compression != null) {
+              builder.addSerde(
+                  DictionaryEncodedColumnPartSerde.createCompressedSingleValue(
```

Comment:
I think a builder pattern would probably end up a bit cleaner.  Essentially the difference I'm seeing is just the `VSizeIntsIndexed`, in which case it would probably be really nice to be able to specify all the other things once and then go into this logic for just specifying the `VSizeIntsIndexed`